test/test_symbolic.py

```diff
@@ -0,0 +1,120 @@
+__copyright__ = """Copyright (C) 2021 University of Illinois Board of Trustees"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+import pyopencl.array as cla  # noqa
+import pyopencl.clmath as clmath # noqa
+from pytools.obj_array import make_obj_array
+import pymbolic as pmbl
+import pymbolic.primitives as prim
+from meshmode.dof_array import thaw
+from meshmode.mesh.generation import generate_regular_rect_mesh
+import mirgecom.symbolic as sym
+
+from meshmode.array_context import (  # noqa
+    pytest_generate_tests_for_pyopencl_array_context
+    as pytest_generate_tests)
+
+import pytest
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.parametrize(("sym_f", "expected_sym_df"), [
+    (1, 0),
+    (pmbl.var("x")**2, 2*pmbl.var("x")),
+    (pmbl.var("cos")(2*pmbl.var("x")), -pmbl.var("sin")(2*pmbl.var("x"))*2),
+    (pmbl.var("sin")(2*pmbl.var("x")), pmbl.var("cos")(2*pmbl.var("x"))*2),
+    (pmbl.var("exp")(2*pmbl.var("x")), pmbl.var("exp")(2*pmbl.var("x"))*2),
+])
+def test_symbolic_diff(sym_f, expected_sym_df):
```

Comment:
```suggestion
def test_symbolic_diff(sym_f, expected_sym_df):
    r"""Demonstrate symbolic differentiation.

    This test demonstrates how to use symbolic differentiation by
    testing that it get the expected results.
    
    Function    Result
    ------------------
    1              0
    $x^2$         $2x$
    $\cos(2x)$    $-2\sin(2x)$
    $\sin(2x)$    $2\cos(2x)$
    $e^{(2x)}$    $2e^{(2x)}$
    """"
```

Reply:
Feel free to ignore this suggestion if you hate it.  For me, a great deal of hand-holding comments would be quite helpful in understanding how to use this. The most confusing parts are those where the constructors for the symbolic vectors are appearing compactly inside the test parameterization.